src/mono/mono/metadata/class-init.c

```diff
@@ -2217,6 +2271,16 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 				}
 
 				size = mono_type_size (field->type, &align);
+				if (m_class_is_inlinearray (klass)) {
+					// limit the max size of array instance to 1MiB
+					const guint32 struct_max_size = 1024 * 1024;
+					size *= m_class_inlinearray_value (klass);
+					inlined_fields++;
+					if(size <= 0 || size > struct_max_size) {
+						mono_class_set_type_load_failure (klass, "Inline array struct size out of bounds.");
```

Comment:
Is this an error that we share with coreclr?  Should we mention the maximum size?

Reply:
> Should we mention the maximum size?

I am not sure the limit is very informative to the user.

The size is computed from element size and will be impacted by aligning and padding. The limit was rather arbitrarily chosen as there would be some limit anyways. 
1 MiB seems to be enough for most uses. It can be increased if there is a need for more, but likely this error will be seen only in erroneous cases.  

We basically want to say "Hey, the array is abnormally large. Is everything alright with the code?"

Reply:
It might be valuable for users to provide more information on the error, either maximum size or indication that the struct is abnormally large. Would it be wrong to mention maximum size as it could be impacted by aligning and padding?